File: RB101/exercises/advanced_1/merge.rb
# given two arrays. Combine the two arrays in a sorted order and return as a single array or sorted elements

# Rules:
# cannot sort the array
#   - must build the result array one element at a time in proper order
# cannot mutate the input arrays
# the two arrays are already sorted

# input: two arrays of integers
# output: merged and sorted array

# Algorithm:
# 1. create an empty array and identifying the current lowest number in either array that is not currently in the result array, count the occurences of the number in either array push those numbers into the next array and move to the next number. When all of the numbers are in the result array return.

# 2. track the indexes for both arrays. at current index, compare the two elements. if element at arr1_idx is less than arr2_idx, push arr1_idx element into array and increase arr1_idx by 1. Do the same for arr2_idx in the opposite instance. repeated until arr1_idx or arr2_idx is == to the size of associated array at which point replace fill the result array with any remaining digits.
# We'll do all of this untill the reuslts array is equal to the size fo the two combined arrays.



# Given two input arrays
# set an empty result array
# set arr1_idx to 0
# set arr2_idx to 0
# until the result array is the size of the two arrays
#   if arr1idx isn't equal to arr1 size and arr2idx != to arr2 size
#     if current arr1 element is less than array2 element
#       push arr1 element into result array
#       add 1 to arr1_idx
#     elsif arr2 element is less than array 1 element
#       push arr2 element into result array
#       add 1 to arr2_idx
#     else
#       push arr1 element and arr2 element into results
#       add 1 to both idx
#     end
#   elsif arr1 is less than array1 size and arr 2 is == arr2size
#     push arr1 element into result
#     add 1 to arr1 idx
#   else
#     push arr2 element into result
#     add 1 to arr2 idx
#   end
# end
# return the results array

# def add_element(array, index)

# end

# def merge(arr1, arr2)
#   result = []
#   idx1= 0
#   idx2 = 0
#   until result.size == arr1.size + arr2.size
#     if idx1 < arr1.size && idx2 < arr2.size
#       if arr1[idx1] < arr2[idx2]
#         result << arr1[idx1]
#         idx1 += 1
#       elsif arr2[idx2] < arr1[idx1]
#         result << arr2[idx2]
#         idx2 += 1
#       else
#         result << arr1[idx1]
#         result << arr2[idx2]
#         idx1 += 1
#         idx2 += 1
#       end
#     elsif idx1 < arr1.size && idx2 == arr2.size
#       result << arr1[idx1]
#       idx1 += 1
#     else
#       result << arr2[idx2]
#       idx2 += 1
#     end
#   end
#   result
# end

def merge(arr1, arr2)
  result = []
  arr1_dup = arr1.dup
  arr2_dup = arr2.dup
  until result.size == arr1.size + arr2.size
    if !arr1_dup.empty? && !arr2_dup.empty?
      if arr1_dup.first < arr2_dup.first
        result << arr1_dup.shift
      else
        result << arr2_dup.shift
      end
    elsif arr1_dup.empty?
      result << arr2_dup.shift
    else
      result << arr1_dup.shift
    end
  end
  result
end 

p merge([1, 5, 9], [2, 6, 8]) == [1, 2, 5, 6, 8, 9]
p merge([1, 1, 3], [2, 2]) == [1, 1, 2, 2, 3]
p merge([], [1, 4, 5]) == [1, 4, 5]
p merge([1, 4, 5], []) == [1, 4, 5]

File: LS215/exercises/medium2/merge.js
/*
  build a new array from two given arrays that contains all of the elements, sorted in ascending order.
  don't mutate the input arrays
  don't use any sorting method on the result array. Place the elements in the new array in order, one at a time
  the arrays provided are already sorted 

use indexes to track the current position while iterating through each array.
compare the two elements at each index and input the smaller value into the new array then incremenet that arrays index
*/

function merge(arr1, arr2) {
  let idx1 = 0;
  let idx2 = 0;
  let result = [];
  let length = arr1.length + arr2.length;
  
  while (result.length < length) {
    arr1El = arr1[idx1];
    arr2El = arr2[idx2];

    if (arr1El <= arr2El || idx2 >= arr2.length) {
      result.push(arr1El);
      idx1 += 1;
    } else if (arr1El > arr2El || idx1 >= arr1.length) {
      result.push(arr2El);
      idx2 += 1;
    }
  }

  console.log(result);
  return result;
}

// function merge(array1, array2) {
//   const copy1 = array1.slice();
//   const copy2 = array2.slice();
//   const result = [];

//   while (copy1.length > 0 && copy2.length > 0) {
//     result.push(copy1[0] <= copy2[0] ? copy1.shift() : copy2.shift());
//   }

//   return result.concat(copy1.length === 0 ? copy2 : copy1);
// }

function mergeSort(array) {
  if (array.length === 1) {
    return array;
  }

  let subArray1 = array.slice(0, array.length / 2);
  let subArray2 = array.slice(array.length / 2);

  subArray1 = mergeSort(subArray1);
  subArray2 = mergeSort(subArray2);

  return merge(subArray1, subArray2);
}

// merge([1, 5, 9], [2, 6, 8]);      // [1, 2, 5, 6, 8, 9]
// merge([1, 1, 3], [2, 2]);         // [1, 1, 2, 2, 3]
// merge([], [1, 4, 5]);             // [1, 4, 5]
// merge([1, 4, 5], []);             // [1, 4, 5]


console.log(mergeSort([9, 5, 7, 1]));           // [1, 5, 7, 9]
console.log(mergeSort([5, 3]));                 // [3, 5]
console.log(mergeSort([6, 2, 7, 1, 4]));        // [1, 2, 4, 6, 7]

console.log(mergeSort(['Sue', 'Pete', 'Alice', 'Tyler', 'Rachel', 'Kim', 'Bonnie']));
// ["Alice", "Bonnie", "Kim", "Pete", "Rachel", "Sue", "Tyler"]

console.log(mergeSort([7, 3, 9, 15, 23, 1, 6, 51, 22, 37, 54, 43, 5, 25, 35, 18, 46]));
// [1, 3, 5, 6, 7, 9, 15, 18, 22, 23, 25, 35, 37, 43, 46, 51, 54]